### PR TITLE
CA-209502: Exception while deleting multiple folders

### PR DESCRIPTION
**Pull request by MihaelaStoica**
- Changed the FolderAction class so that it can be used to delete or move multiple folders across connections, instead of having multiple actions (one for each holder) which can not run in parallel.
- The api calls that update the EMPTY_FOLDERS entry in pool's other_config are preformed last for all connections, avoiding cross-thread exceptions when two threads try to add/remove the same key.

Signed-off-by: Mihaela Stoica <mihaela.stoica@citrix.com>

### CODE REVIEW

File: XenModel/Actions/Folders/FolderAction.cs [hunk 1]
```diff
@@ -192,9 +227,68 @@ public override void RecomputeCanCancel()
...
+            {
+                AddFoldersToEmptyList(con, cancelling, emptyFolders[con].ToArray());
+            }
+            RemoveFoldersFromEmptyList(nonEmptyFolders.ToArray(), cancelling);
+        }
+
+        internal void DeleteOrMove(List<IXenObject> objects, Folder folder, Func<bool> cancelling)
+        {
+            IXenConnection connection = null;
+            if (objects.Count > 0 && objects[0] != null)
```
Comment (kc284) [inline XenModel/Actions/Folders/FolderAction.cs]: Is it likely to have null objects in the list? If so we need to add a null check in the other DeleteOrMove overload.
Alternatively we can filter out null objects at the beginning of this method - also return if the list is empty.

**Comment by kc284** [comment]
Apart from the comment above, the changes look ok to me.

Not strictly related to this ticket, but overall the FolderAction could use some subclassing to get rid of the Kind enum, simplify the code and be more object oriented instead of having comments on the use of constructors etc.

File: XenModel/Actions/Folders/FolderAction.cs (hunk 1 above)
Comment (MihaelaStoica) [inline XenModel/Actions/Folders/FolderAction.cs]: The null check is not needed, we should never have null objects in the list.

**Comment by MihaelaStoica** [comment]
Thanks @kc284 for the suggestion. I've refactored the FolderAction file, with subclasses for each action kind.

**Comment by kc284** [comment]
Look better now. Merging.

**Pull request merged**